remote.go

```diff
@@ -0,0 +1,265 @@
+package maddy
+
+import (
+	"bytes"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/textproto"
+	"sort"
+	"strings"
+
+	"github.com/emersion/go-smtp"
+	"github.com/emersion/maddy/config"
+	"github.com/emersion/maddy/log"
+	"github.com/emersion/maddy/module"
+)
+
+var ErrTLSRequired = errors.New("TLS is required for outgoing connections but target server doesn't support STARTTLS")
+
+type RemoteDelivery struct {
+	name       string
+	hostname   string
+	requireTLS bool
+
+	Log log.Logger
+}
+
+func NewRemoteDelivery(_, instName string) (module.Module, error) {
+	return &RemoteDelivery{
+		name: instName,
+		Log:  log.Logger{Name: "remote"},
+	}, nil
+}
+
+func (rd *RemoteDelivery) Init(cfg *config.Map) error {
+	cfg.String("hostname", true, true, "", &rd.hostname)
+	cfg.Bool("debug", true, &rd.Log.Debug)
+	cfg.Bool("require_tls", false, &rd.requireTLS)
+	if _, err := cfg.Process(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rd *RemoteDelivery) Name() string {
+	return "remote"
+}
+
+func (rd *RemoteDelivery) InstanceName() string {
+	return rd.name
+}
+
+func (rd *RemoteDelivery) Deliver(ctx module.DeliveryContext, msg io.Reader) error {
+	var body io.ReadSeeker
+	if seekable, ok := msg.(io.ReadSeeker); ok {
+		body = seekable
+	} else {
+		bodySlice, err := ioutil.ReadAll(msg)
+		if err != nil {
+			return errors.New("failed to buffer message")
+		}
+		body = bytes.NewReader(bodySlice)
+	}
+
+	if _, err := body.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	groupedRcpts, err := groupByDomain(ctx.To)
+	if err != nil {
+		return err
+	}
+
+	partialErr := PartialError{
+		Errs: make(map[string]error),
+	}
+	// TODO: look into ways to parallelize this, the main trouble here is body
+	// probably create pipe for each server and copy body to each?
```

Comment:
Could also call `bytes.NewReader` for each server. We should be careful not to completely stall everything if one server is slow.

Reply:
Should I reimplement it using `bytes.NewReader` or leave it as is for now?

Reply:
I'm fine with the current behaviour, I prefer adding features that are simple and work and then improve them in future PRs.